### PR TITLE
Document requirements cache behavior differences.

### PR DESCRIPTION
**Pull request by tvalentyn**

### CODE REVIEW

**Comment by tvalentyn** [comment]
R: @rszper

File: website/www/site/content/en/documentation/sdks/python-pipeline-dependencies.md [hunk 1]
```diff
@@ -46,11 +46,21 @@ To supply a requirements.txt file:
 
     The runner will use the `requirements.txt` file to install your additional dependencies onto the remote workers.
 
-> **NOTE**: An alternative to `pip freeze` is to use a library like [pip-tools](https://github.com/jazzband/pip-tools) to compile all the dependencies required for the pipeline from a `--requirements_file`, where only top-level dependencies are mentioned.
+> **NOTE**: An alternative to `pip freeze` is to use a library like [pip-tools](https://github.com/jazzband/pip-tools) to compile all the dependencies required for the pipeline from a `requirements.in` file, where only the top-level dependencies are mentioned.
```
Comment (rszper) [inline website/www/site/content/en/documentation/sdks/python-pipeline-dependencies.md:49]:
```suggestion
> **NOTE**: As an alternative to `pip freeze`, use a library like [pip-tools](https://github.com/jazzband/pip-tools) to compile all of the dependencies required for the pipeline from a `requirements.in` file. In the `requirements.in` file, only the top-level dependencies are mentioned.
```

File: website/www/site/content/en/documentation/sdks/python-pipeline-dependencies.md [hunk 2]
```diff
@@ -46,11 +46,21 @@ To supply a requirements.txt file:
...
+
+When you supply the `--requirements_file` pipeline option, Beam downloads
+specified packages locally into a requirements cache directory during pipeline
```
Comment (rszper) [inline website/www/site/content/en/documentation/sdks/python-pipeline-dependencies.md:52]:
```suggestion
the specified packages locally into a requirements cache directory,
```

File: website/www/site/content/en/documentation/sdks/python-pipeline-dependencies.md [hunk 3]
```diff
@@ -46,11 +46,21 @@ To supply a requirements.txt file:
 
     The runner will use the `requirements.txt` file to install your additional dependencies onto the remote workers.
 
-> **NOTE**: An alternative to `pip freeze` is to use a library like [pip-tools](https://github.com/jazzband/pip-tools) to compile all the dependencies required for the pipeline from a `--requirements_file`, where only top-level dependencies are mentioned.
+> **NOTE**: An alternative to `pip freeze` is to use a library like [pip-tools](https://github.com/jazzband/pip-tools) to compile all the dependencies required for the pipeline from a `requirements.in` file, where only the top-level dependencies are mentioned.
+
+When you supply the `--requirements_file` pipeline option, Beam downloads
```
Comment (rszper) [inline website/www/site/content/en/documentation/sdks/python-pipeline-dependencies.md:51]:
```suggestion
When you supply the `--requirements_file` pipeline option,  during pipeline submission, Beam downloads
```

File: website/www/site/content/en/documentation/sdks/python-pipeline-dependencies.md [hunk 4]
```diff
@@ -46,11 +46,21 @@ To supply a requirements.txt file:
...
+specified packages locally into a requirements cache directory during pipeline
+submission, and stages the requirements cache directory to the runner.
```
Comment (rszper) [inline website/www/site/content/en/documentation/sdks/python-pipeline-dependencies.md:53]:
```suggestion
and then stages the requirements cache directory to the runner.
```

File: website/www/site/content/en/documentation/sdks/python-pipeline-dependencies.md [hunk 5]
```diff
@@ -46,11 +46,21 @@ To supply a requirements.txt file:
...
+At pipeline runtime, Beam prefers to install packages from requirements cache
```
Comment (rszper) [inline website/www/site/content/en/documentation/sdks/python-pipeline-dependencies.md:54]:
```suggestion
At runtime, when available, Beam installs packages from the requirements cache.
```

File: website/www/site/content/en/documentation/sdks/python-pipeline-dependencies.md [hunk 6]
```diff
@@ -46,11 +46,21 @@ To supply a requirements.txt file:
...
+if available. This mechanism allows staging dependency packages to the runner
```
Comment (rszper) [inline website/www/site/content/en/documentation/sdks/python-pipeline-dependencies.md:55]:
```suggestion
This mechanism makes it possible to stage the dependency packages to the runner
```

File: website/www/site/content/en/documentation/sdks/python-pipeline-dependencies.md [hunk 7]
```diff
@@ -46,11 +46,21 @@ To supply a requirements.txt file:
...
+at submission, and at runtime the runner workers might be able to install the
```
Comment (rszper) [inline website/www/site/content/en/documentation/sdks/python-pipeline-dependencies.md:56]:
```suggestion
at submission. At runtime, the runner workers might be able to install the
```

File: website/www/site/content/en/documentation/sdks/python-pipeline-dependencies.md [hunk 8]
```diff
@@ -46,11 +46,21 @@ To supply a requirements.txt file:
...
+packages from cache, without a connection to PyPI. To disable staging the
```
Comment (rszper) [inline website/www/site/content/en/documentation/sdks/python-pipeline-dependencies.md:57]:
```suggestion
packages from the cache without needing a connection to PyPI. To disable staging the
```

File: website/www/site/content/en/documentation/sdks/python-pipeline-dependencies.md [hunk 9]
```diff
@@ -46,11 +46,21 @@ To supply a requirements.txt file:
...
+requirements, supply the `--requirements_cache=skip` pipeline option.
```
Comment (rszper) [inline website/www/site/content/en/documentation/sdks/python-pipeline-dependencies.md:58]:
```suggestion
requirements, use the `--requirements_cache=skip` pipeline option.
```

File: website/www/site/content/en/documentation/sdks/python-pipeline-dependencies.md [hunk 10]
```diff
@@ -46,11 +46,21 @@ To supply a requirements.txt file:
...
+For more information, see the [help descriptions of these pipeline options](https://beam.apache.org/releases/pydoc/current/_modules/apache_beam/options/pipeline_options.html#SetupOptions).
 
 ## Custom Containers {#custom-containers}
 
-You can pass a [container](https://hub.docker.com/search?q=apache%2Fbeam&type=image) image with all the dependencies that are needed for the pipeline instead of `requirements.txt`. [Follow the instructions on how to run pipeline with Custom Container images](/documentation/runtime/environments/#running-pipelines).
+You can pass a [container](https://hub.docker.com/search?q=apache%2Fbeam&type=image) image with all the dependencies that are needed for the pipeline. [Follow the instructions on how to run pipeline with Custom Container images](/documentation/runtime/environments/#running-pipelines).
```
Comment (rszper) [inline website/www/site/content/en/documentation/sdks/python-pipeline-dependencies.md:63]:
```suggestion
You can pass a [container](https://hub.docker.com/search?q=apache%2Fbeam&type=image) image with all the dependencies that are needed for the pipeline. [Follow the instructions the show how to run the pipeline with custom container images](/documentation/runtime/environments/#running-pipelines).
```

File: website/www/site/content/en/documentation/sdks/python-pipeline-dependencies.md [hunk 11]
```diff
@@ -118,7 +128,10 @@ Often, your pipeline code spans multiple files. To run your project remotely, yo
 
         --setup_file /path/to/setup.py
 
-**Note:** If you [created a requirements.txt file](#pypi-dependencies) and your project spans multiple files, you can get rid of the `requirements.txt` file and instead, add all packages contained in `requirements.txt` to the `install_requires` field of the setup call (in step 1).
+**Note:** It is not necessary to supply the `--requirements_file` [option](#pypi-dependencies) if the dependenices of your package are defined in the `install_requires` field of the `setup.py` file (see step 1).
```
Comment (rszper) [inline website/www/site/content/en/documentation/sdks/python-pipeline-dependencies.md:131]:
```suggestion
**Note:** It is not necessary to supply the `--requirements_file` [option](#pypi-dependencies) if the dependencies of your package are defined in the `install_requires` field of the `setup.py` file (see step 1).
```

File: website/www/site/content/en/documentation/sdks/python-pipeline-dependencies.md [hunk 12]
```diff
@@ -118,7 +128,10 @@ Often, your pipeline code spans multiple files. To run your project remotely, yo
...
+However unlike the `--requirements_file` option, when  using the `--setup_file` option, Beam does not stage the dependent packages to the Runner,
```
Comment (rszper) [inline website/www/site/content/en/documentation/sdks/python-pipeline-dependencies.md:132]:
```suggestion
However unlike with the `--requirements_file` option, when you use the `--setup_file` option, Beam doesn't stage the dependent packages to the runner.
```

File: website/www/site/content/en/documentation/sdks/python-pipeline-dependencies.md [hunk 13]
```diff
@@ -118,7 +128,10 @@ Often, your pipeline code spans multiple files. To run your project remotely, yo
...
+only the pipeline package is staged and its dependencies are installed from PyPI
```
Comment (rszper) [inline website/www/site/content/en/documentation/sdks/python-pipeline-dependencies.md:133]:
```suggestion
Only the pipeline package is staged. If they aren't already provided in the runtime environment,
```

File: website/www/site/content/en/documentation/sdks/python-pipeline-dependencies.md [hunk 14]
```diff
@@ -118,7 +128,10 @@ Often, your pipeline code spans multiple files. To run your project remotely, yo
...
+at runtime if not already provided in the runtime environment.
```
Comment (rszper) [inline website/www/site/content/en/documentation/sdks/python-pipeline-dependencies.md:134]:
```suggestion
the package dependencies are installed from PyPI at runtime.
```

**Review by rszper — APPROVED** [review]

**Comment by tvalentyn** [comment]
Thank you!

**Pull request merged**